components/tidb_query_datatype/src/codec/convert.rs

```diff
@@ -496,7 +496,9 @@ impl<'a> ToInt for JsonRef<'a> {
         // TiDB:  5
         // MySQL: 4
         let val = match self.get_type() {
-            JsonType::Object | JsonType::Array => Ok(0),
+            JsonType::Object | JsonType::Array => Ok(ctx
+                .handle_truncate_err(Error::incorrect_datetime_value("Integer"))
```

Comment:
Maybe we need to introduce a new error type instead of  `truncate_err` and `incorrect_datetime_value`?

Reply:
It's my careless mistake to use "incorrect_datetime_value" err.
To keep up with tidb implementation, fixed it with truncated wrong value error, which err msg will be like: `Data truncation: Truncated incorrect INTEGER value: '{"obj": 100}'`.